nipoppy/workflows/pipeline.py

```diff
@@ -39,6 +40,36 @@
 from nipoppy.workflows.base import BaseWorkflow
 
 
+def apply_analysis_level(
+    participants_sessions: Iterable[str, str],
+    analysis_level: AnalysisLevelType,
+) -> Tuple[str, str]:
+    """Filter participant-session pairs to run based on the analysis level."""
+    if analysis_level == AnalysisLevelType.group:
+        participants_session_updated = [(None, None)]
+
+    elif analysis_level == AnalysisLevelType.participant:
+        participants = []
+        for participant, _ in participants_sessions:
+            if participant not in participants:
+                participants.append(participant)
+        participants_session_updated = [
+            (participant, None) for participant in participants
+        ]
+
+    elif analysis_level == AnalysisLevelType.session:
+        sessions = []
+        for _, session in participants_sessions:
+            if session not in sessions:
+                sessions.append(session)
+        participants_session_updated = [(None, session) for session in sessions]
+
+    else:
+        participants_session_updated = participants_sessions
+
+    return participants_session_updated
```

Comment:
I find it easier to read when the return statement is within the condition (see below). It avoids having to read the rest of the code to see if something modify the variable. Although, I don't know what the best practice is.

```python3
def apply_analysis_level(
    participants_sessions: Iterable[str, str],
    analysis_level: AnalysisLevelType,
) -> Tuple[str, str]:
  """Filter participant-session pairs to run based on the analysis level."""
  if analysis_level == AnalysisLevelType.group:
       return [(None, None)]

    elif analysis_level == AnalysisLevelType.participant:
        participants = []
        for participant, _ in participants_sessions:
            if participant not in participants:
                participants.append(participant)
        return [(participant, None) for participant in participants]

    elif analysis_level == AnalysisLevelType.session:
        sessions = []
        for _, session in participants_sessions:
            if session not in sessions:
                sessions.append(session)
        return [(None, session) for session in sessions]

    else:
        return participants_sessions

```